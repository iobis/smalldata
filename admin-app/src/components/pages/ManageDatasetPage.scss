.manage-dataset-page {
  .datasets {
    margin-top: var(--margin-top);
  }
}
