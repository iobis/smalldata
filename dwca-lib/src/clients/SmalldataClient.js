import { mapDatasetToRequest, mapOccurrenceToDwca } from './SmalldataConverters'

const authorizationValue = 'Basic verysecret'
const headers = {
  'Authorization': authorizationValue,
  'Content-Type':  'application/json'
}

export async function getDatasets() {
  const response = await fetch('/api/datasets', { headers })
    .then(response => response.json())
  return response.map(renameRefToId)
}

export async function createDataset(dataset) {
  const request = mapDatasetToRequest(dataset)
  return await fetch('/api/datasets', {
    method: 'POST',
    headers,
    body:   JSON.stringify(request)
  }).then(response => response.json())
}

export function renameRefToId({ ref, ...rest }) {
  return ({ id: ref, ...rest })
}

export async function getOccurrences({ userRef }) {
  const query = 'projectFields=dwcRecord.tdwg.scientificName&projectFields=dwcRecord.tdwg.eventDate'
  return await fetch(`/api/dwca/user/${userRef}/records?${query}`, { headers })
    .then(response => response.json())
}

export async function getOccurrence({ datasetId, dwcaId, userRef }) {
  return fetch(`/api/dwca/${datasetId}/user/${userRef}/records/${encodeURIComponent(dwcaId)}`, { headers })
    .then(response => response.json())
}

export async function findLatestOccurrence({ userRef }) {
  const occurrences = await getOccurrences({ userRef })
  const latest = occurrences
    .filter(dwca => !!dwca.addedAtInstant)
    .sort((left, right) => left.addedAtInstant < right.addedAtInstant ? 1 : -1)[0]
  return await getOccurrence({ userRef, dwcaId: latest.dwcaId, datasetId: latest.dataset })
}

export async function updateOccurrence({ occurrence, userRef, dwcaId }) {
  const request = mapOccurrenceToDwca(occurrence)
  const url = `/api/dwca/${occurrence.dataset.id}/user/${userRef}/records/${encodeURIComponent(dwcaId)}`
  return await fetch(url, {
    method: 'PUT',
    headers,
    body:   JSON.stringify(request)
  }).then(response => response.json())
}

export async function createOccurrence({ occurrence, userRef }) {
  const request = mapOccurrenceToDwca(occurrence)
  const url = `/api/dwca/${occurrence.dataset.id}/user/${userRef}/records`
  return await fetch(url, {
    method: 'POST',
    headers,
    body:   JSON.stringify(request)
  }).then(response => response.json())
}

export async function getUsers() {
  return fetch('/api/users', { headers })
    .then(response => response.json())
}

export async function getUsersWithDatasets() {
  const [users, datasets] = await Promise.all([getUsers(), getDatasets()])
  const datasetIdToDataset = groupBy(datasets, 'id')
  return users.map(user => {
    if (!user['dataset_refs']) user['dataset_refs'] = []
    const datasets = user['dataset_refs'].map(datasetRef => datasetIdToDataset[datasetRef][0])
    return { ...user, datasets }
  })
}

export async function createUser({ datasetIds, email, name, role }) {
  const request = {
    'dataset_refs': datasetIds,
    emailAddress:   email,
    name,
    role
  }
  const url = '/api/users'
  return await fetch(url, {
    method: 'POST',
    headers,
    body:   JSON.stringify(request)
  }).then(response => response.json())
    .catch(error => ({
      exception:        'Some error occurred',
      exceptionMessage: error
    }))
}

export async function updateUser({ id, datasetIds, email, name, role }) {
  const request = {
    'dataset_refs': datasetIds,
    emailAddress:   email,
    name,
    role
  }
  const url = '/api/users/' + id
  return await fetch(url, {
    method: 'PUT',
    headers,
    body:   JSON.stringify(request)
  }).then(response => response.json())
    .catch(error => ({
      exception:        'Some error occurred',
      exceptionMessage: error
    }))
}

function groupBy(list, props) {
  return list.reduce((a, b) => {
    (a[b[props]] = a[b[props]] || []).push(b)
    return a
  }, {})
}
