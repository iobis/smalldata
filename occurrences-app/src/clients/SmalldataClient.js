import { format } from 'date-fns'

export async function getDatasets() {
  const response = await fetch('/api/datasets')
    .then(response => response.json())
  return response.map(renameRefToId)
}

export function renameRefToId({ ref, ...rest }) {
  return ({ id: ref, ...rest })
}

export function datasetTitleOf(dataset) {
  const title = dataset && dataset.title && dataset.title.value
  return title || ''
}

export function getOccurrenceMock() {
  return [{
    id:             1,
    addedDate:      '2011-12-01',
    scientificName: 'Abra Alba',
    dataset:        'NPPSD Short-tailed Albatross Sightings',
    occurrenceDate: '2011-12-09'
  }, {
    id:             2,
    addedDate:      '2011-12-01',
    scientificName: 'Abra Alba',
    dataset:        'NPPSD Short-tailed Albatross Sightings',
    occurrenceDate: '2011-12-09'
  }]
}

export async function postOccurrence({ occurence }) {
  const userRef = 'ovZTtaOJZ98xDDY'
  const occurrence = {
    core:       'occurrence',
    occurrence: [{
      tdwg: {
        datasetID:                     occurence.dataset.id,
        datasetName:                   occurence.dataset.title.value,
        basisOfRecord:                 occurence.occurrenceData.basisOfRecord.charAt(0).toUpperCase() + occurence.occurrenceData.basisOfRecord.slice(1),
        eventDate:                     format(occurence.occurrenceData.beginDate, 'YYYY-MM-DD') + '/' + format(occurence.occurrenceData.endDate, 'YYYY-MM-DD'),
        lifestage:                     occurence.occurrenceData.lifeStage,
        occurrenceStatus:              occurence.occurrenceData.occurrenceStatus,
        scientificName:                occurence.occurrenceData.scientificName,
        sex:                           occurence.occurrenceData.sex,
        decimalLongitude:              occurence.locationData.decimalLongitude,
        decimalLatitude:               occurence.locationData.decimalLatitude,
        coordinateUncertaintyInMeters: occurence.locationData.coordinateUncertainty,
        minimumDepthInMeters:          occurence.locationData.minimumDepth,
        maximumDepthInMeters:          occurence.locationData.maximumDepth,
        verbatimCoordinates:           occurence.locationData.verbatimCoordinates,
        verbatimDepth:                 occurence.locationData.verbatimDepth
      }
    }],
    emof:       [{
      purl:  {},
      iobis: {}
    }, {
      iobis: {}
    }]
  }
  const url = `/api/dwca/${occurence.dataset.id}/user/${userRef}/records`
  return await fetch(url, {
    method:  'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body:    JSON.stringify(occurrence)
  }).then(response => response.json())
}
