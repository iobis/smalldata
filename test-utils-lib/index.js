export function ignoreActWarning() {
  const originalError = console.error

  beforeAll(() => {
    console.error = (...args) => {
      if (/Warning.*not wrapped in act/.test(args[0])) return
      originalError.call(console, ...args)
    }
  })

  afterAll(() => {
    console.error = originalError
  })
}

export function flushPromises() {
  return new Promise(resolve => setImmediate(resolve))
}
