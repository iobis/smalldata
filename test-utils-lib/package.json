{
  "name": "@smalldata/test-utils-lib",
  "version": "1.0.0",
  "private": true,
  "main": "index.js",
  "scripts": {
    "lint": "echo Nothig to lint yet",
    "test": "echo Nothig to test here",
    "test:ci": "echo Nothig to test here"
  }
}
