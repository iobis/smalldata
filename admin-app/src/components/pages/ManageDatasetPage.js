import React, { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'
import Divider from '../layout/Divider'
import { getDatasets } from '@smalldata/dwca-lib/src/clients/SmalldataClient'

export default function ManageDatasetPage() {
  const { t } = useTranslation()
  const [datasets, setDatasets] = useState([])

  useEffect(() => {
    const fetchDatasets = async() => {
      setDatasets(await getDatasets())
    }
    fetchDatasets()
  }, [])

  return (
    <>
      <section className="section">
        <div className="container is-fluid has-text-centered">
          <Link className="button is-info" to="/manage-dataset/create">
            {t('manageDatasetPage.inputNew')}
          </Link>
        </div>
      </section>
      <Divider>{t('manageDatasetPage.or')}</Divider>
      <section className="section">
        <div className="container is-fluid has-text-centered">
          <h4 className="title is-4">{t('manageDatasetPage.managePreviousEntries')}</h4>
          <table className="table is-striped is-hoverable is-fullwidth">
            <thead>
              <tr>
                <th/>
                <th>{t('manageDatasetPage.table.title')}</th>
                <th>{t('manageDatasetPage.table.organization')}</th>
                <th>{t('manageDatasetPage.table.licence')}</th>
              </tr>
            </thead>
            <tbody>
              {datasets.map(dataset => (
                <DatasetRow
                  key={dataset.id}
                  license={dataset.license.title}
                  organization={dataset.metadataProviders[0].organizationName || 'â€”'}
                  title={dataset.title.value}/>
              ))}
            </tbody>
          </table>
        </div>
      </section>
    </>
  )
}

function DatasetRow({ title, organization, license }) {
  const { t } = useTranslation()

  const toEdit = {
    pathname: '/manage-dataset/update',
    state:    { action: 'update' }
  }

  return (
    <tr className="user-row">
      <td className="edit">
        <Link className="button is-info" to={toEdit}>
          {t('common.edit')}
        </Link>
      </td>
      <td className="dataset-title">{title}</td>
      <td className="organization">{organization}</td>
      <td className="license">{license}</td>
    </tr>
  )
}

DatasetRow.propTypes = {
  license:      PropTypes.string.isRequired,
  organization: PropTypes.string.isRequired,
  title:        PropTypes.string.isRequired
}
