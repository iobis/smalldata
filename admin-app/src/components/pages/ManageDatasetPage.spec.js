import SmalldataClient from '@smalldata/dwca-lib/src/clients/SmalldataClient'
import { flushPromises, ignoreActWarning } from '@smalldata/test-utils-lib'
import ManageDatasetPage from './ManageDatasetPage'
import { MemoryRouter } from 'react-router-dom'
import { mount } from 'enzyme'
import React from 'react'
import { act } from 'react-dom/test-utils'

jest.mock('@smalldata/dwca-lib/src/clients/SmalldataClient', () => ({
  getDatasets: jest.fn().mockImplementation(() =>
    new Promise((resolve) => {
      resolve([{
        id:                'id-1',
        license:           {
          title: 'title-1'
        },
        metadataProviders: [{
          organizationName: 'Flanders Marine Institute (VLIZ)'
        }],
        title:             {
          value: 'with organization name in metadata'
        }
      }, {
        id:                'id-2',
        license:           {
          title: 'title-2'
        },
        metadataProviders: [{}],
        title:             {
          value: 'without organization name in metadata'
        }
      }])
    })
  )
}))

describe('ManageDatasetPage', () => {
  ignoreActWarning()

  let wrapper

  it('renders correctly for non empty datasets', async() => {
    act(() => {
      wrapper = mount(
        <MemoryRouter initialEntries={[{ pathname: '/manage-dataset', key: 'testKey' }]}>
          <ManageDatasetPage/>
        </MemoryRouter>
      )
    })
    expect(SmalldataClient.getDatasets).toHaveBeenCalledWith()

    await flushPromises()
    wrapper.update()
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('tbody tr')).toHaveLength(2)
  })
})
