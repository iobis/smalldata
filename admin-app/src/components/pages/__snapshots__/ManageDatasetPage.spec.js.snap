// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageDatasetPage renders correctly for non empty datasets 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
        "pathname": "/manage-dataset",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/manage-dataset",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/manage-dataset",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <ManageDatasetPage>
      <section
        className="section"
      >
        <div
          className="container is-fluid has-text-centered"
        >
          <Link
            className="button is-info"
            to="/manage-dataset/create"
          >
            <a
              className="button is-info"
              href="/manage-dataset/create"
              onClick={[Function]}
            >
              manageDatasetPage.inputNew
            </a>
          </Link>
        </div>
      </section>
      <Divider>
        <div
          className="divider"
        >
          <span>
            manageDatasetPage.or
          </span>
        </div>
      </Divider>
      <section
        className="section"
      >
        <div
          className="container is-fluid has-text-centered"
        >
          <h4
            className="title is-4"
          >
            manageDatasetPage.managePreviousEntries
          </h4>
          <table
            className="table is-striped is-hoverable is-fullwidth"
          >
            <thead>
              <tr>
                <th />
                <th>
                  manageDatasetPage.table.title
                </th>
                <th>
                  manageDatasetPage.table.organization
                </th>
                <th>
                  manageDatasetPage.table.licence
                </th>
              </tr>
            </thead>
            <tbody>
              <DatasetRow
                key="id-1"
                license="title-1"
                organization="Flanders Marine Institute (VLIZ)"
                title="with organization name in metadata"
              >
                <tr
                  className="user-row"
                >
                  <td
                    className="edit"
                  >
                    <Link
                      className="button is-info"
                      to={
                        Object {
                          "pathname": "/manage-dataset/update",
                          "state": Object {
                            "action": "update",
                          },
                        }
                      }
                    >
                      <a
                        className="button is-info"
                        href="/manage-dataset/update"
                        onClick={[Function]}
                      >
                        common.edit
                      </a>
                    </Link>
                  </td>
                  <td
                    className="dataset-title"
                  >
                    with organization name in metadata
                  </td>
                  <td
                    className="organization"
                  >
                    Flanders Marine Institute (VLIZ)
                  </td>
                  <td
                    className="license"
                  >
                    title-1
                  </td>
                </tr>
              </DatasetRow>
              <DatasetRow
                key="id-2"
                license="title-2"
                organization="—"
                title="without organization name in metadata"
              >
                <tr
                  className="user-row"
                >
                  <td
                    className="edit"
                  >
                    <Link
                      className="button is-info"
                      to={
                        Object {
                          "pathname": "/manage-dataset/update",
                          "state": Object {
                            "action": "update",
                          },
                        }
                      }
                    >
                      <a
                        className="button is-info"
                        href="/manage-dataset/update"
                        onClick={[Function]}
                      >
                        common.edit
                      </a>
                    </Link>
                  </td>
                  <td
                    className="dataset-title"
                  >
                    without organization name in metadata
                  </td>
                  <td
                    className="organization"
                  >
                    —
                  </td>
                  <td
                    className="license"
                  >
                    title-2
                  </td>
                </tr>
              </DatasetRow>
            </tbody>
          </table>
        </div>
      </section>
    </ManageDatasetPage>
  </Router>
</MemoryRouter>
`;
